toolchain/driver/driver.cpp

```diff
@@ -22,286 +25,481 @@
-  } else if (!args.empty() && args[0] == "--print-errors=streamed") {
-    args = args.drop_front();
-  } else {
-    sorting_consumer = std::make_unique<SortingDiagnosticConsumer>(*consumer);
-    consumer = sorting_consumer.get();
+  enum class Phase {
+    Tokenize,
+    Parse,
+    Syntax,
+    Lower,
+    CodeGen,
+  };
+
+  friend auto operator<<(llvm::raw_ostream& out, Phase phase)
+      -> llvm::raw_ostream& {
+    switch (phase) {
+      case Phase::Tokenize:
+        out << "tokenize";
+        break;
+      case Phase::Parse:
+        out << "parse";
+        break;
+      case Phase::Syntax:
+        out << "syntax";
+        break;
+      case Phase::Lower:
+        out << "lower";
+        break;
+      case Phase::CodeGen:
+        out << "codegen";
+        break;
+    }
+    return out;
   }
 
-  if (args.empty()) {
-    error_stream_ << "ERROR: No subcommand specified.\n";
-    return false;
+  void Build(Args::CommandBuilder& b) {
+    b.AddStringPositionalArg(
+        {
+            .name = "FILE",
+            .help = R"""(
+The input Carbon source file to compile.
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.Required(true);
+          arg_b.Set(&input_file_name);
+        });
+
+    b.AddOneOfOption(
+        {
+            .name = "phase",
+            .help = R"""(
+Selects the compilation phase to run. These phases are always run in sequence,
+so every phase before the one selected will also be run. The default is to
+compile, lower, and generate machine code.
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.SetOneOf(
+              {
+                  arg_b.OneOfValue("tokenize", Phase::Tokenize),
+                  arg_b.OneOfValue("parse", Phase::Parse),
+                  arg_b.OneOfValue("syntax", Phase::Syntax),
+                  arg_b.OneOfValue("lower", Phase::Lower),
+                  arg_b.OneOfValue("codegen", Phase::CodeGen).Default(true),
+              },
+              &phase);
+        });
+
+    b.AddStringOption(
+        {
+            .name = "output",
+            .value_name = "FILE",
+            .help = R"""(
+The output filename for codegen.
+
+When this is a file name, either textual assembly or a binary object will be
+written to it based on the flag `--asm-output`. The default is to write a binary
+object file.
+
+This can also be the string `-` which will write the output to stdout. In that
+case, the flag `--asm-output` is ignored and the output defaults to textual
+assembly. Binary object output can be forced by enabling `--force-obj-output`.
+)""",
+        },
+        [&](auto& arg_b) { arg_b.Set(&output_file_name); });
+
+    b.AddStringOption(
+        {
+            .name = "target",
+            .help = R"""(
+Select a target platform. Uses the LLVM target syntax, often called a "triple"
+(despite not actually being a triple in many common cases).
+
+This corresponds to the `target` flag to Clang and accepts the same strings
```

Comment:
```suggestion
Select a target platform. This corresponds to the `target` flag to Clang and accepts the same strings
```

Seems to be saying the same thing both ways, I think the "triple" comments can be elided ("despite not actually being a triple" in particular feels like you're criticizing Clang, and won't help users).

Reply:
Happy to rephrase here to your suggestion or something else.

Just so we're on the same page, some context on what I was trying to accomplish here -- it may still not make sense though, so really am OK with the suggestion (but think this whole thread will vanish if I commit it).

I had wanted to write the docs for the driver more focused on LLVM than Clang given that it is somewhat the common denomenator -- whether C++, Rust, Swift, or some other language as background, it seems reasonable for developers to be thinking about the fact that they're passing essentially LLVM flags here. But maybe that's not the right approach, I can see the value of using Clang as the basis here, especially when we match it very closely. I mean, we're linking to Clang docs here, so yeah, maybe this idea doesn't pan out.

The discussion of triples isn't criticism of Clang really, the terminology (and IMO, the confusion) long predates Clang and comes from autotools AFAIK. Lots of documentation on these strings specifically calls them "triples" or "triplets" and I wanted to avoid folks thinking they had found the wrong docs because it used a term that didn't seem to apply. And that's extra true because I wanted to link them to the Clang docs that call them triples. Maybe there is a better way to achieve this though?

Reply:
```suggestion
Select a target platform. Uses the LLVM target syntax. Also known as a "triple" for historical reasons.

This corresponds to the `target` flag to Clang and accepts the same strings
```


Here's a different suggestion. Versus "despite not actually being a triple in many common cases", "for historical reasons" still implies it's not always a triple, and I think phrasing like that is easier to have interpreted as objective and informative.

Note, if it helps understand how I'm reading this, I think the use of "actually" probably has the strongest influence on my interpretation right now, partly because it's only there for emphasis.